pkg/apiserver/admission/admissiontimeout/timeoutadmission.go

```diff
@@ -29,17 +31,46 @@ func (p pluginHandlerWithTimeout) Admit(ctx context.Context, a admission.Attribu
 		return nil
 	}
 
-	admissionDone := make(chan struct{})
-	admissionErr := fmt.Errorf("default to mutation error")
+	type result struct {
+		admissionErr error
+		panicErr     error
+	}
+	// if a timeout occurs, we don't want the child goroutine to hang forever
+	resultCh := make(chan result, 1)
 	go func() {
-		defer utilruntime.HandleCrash()
-		defer close(admissionDone)
-		admissionErr = mutatingHandler.Admit(ctx, a, o)
+		r := result{}
+		// NOTE: panics don't cross goroutine boundaries, so we have to handle
+		// the error here, we can't call utilruntime.HandleCrash here, then it
+		// will cause the apiserver to crash.
+		// We also need to make sure that the panic is propagated to the caller.
+		defer func() {
+			if err := recover(); err != nil {
+				// do not wrap the sentinel ErrAbortHandler panic value
+				if err != http.ErrAbortHandler {
+					// Same as stdlib http server code. Manually allocate stack
+					// trace buffer size to prevent excessively large logs
+					const size = 64 << 10
+					buf := make([]byte, size)
+					buf = buf[:runtime.Stack(buf, false)]
+					err = fmt.Sprintf("%v\n%s", err, buf)
```

Comment:
This line is just weird to me though it will work. `err` was an error and then now it is a string?